src/diagonal.works/b6/api/functions/collections.go

```diff
@@ -341,3 +342,155 @@ var _ api.Collection = &flattenCollection{}
 func flatten(_ *api.Context, c api.Collection) (api.Collection, error) {
 	return &flattenCollection{c: c}, nil
 }
+
+const maxBuckets = 5
+
+type bucket struct {
+	start, end interface{}
+}
+
+func bucketRange(b bucket) interface{} {
+	if b.start == b.end || b.end == nil {
+		return b.start
+	} else {
+		return fmt.Sprint(b.start) + "-" + fmt.Sprint(b.end)
+	}
+}
+
+type bucketFn func(interface{}, bucket) bool // Indicates whether given value can be bucketed.
+
+func bucketIndex(v interface{}, b []bucket, fn bucketFn) (int, error) {
+	for i, bucket := range b {
+		if fn(v, bucket) {
+			return i, nil
+		}
+	}
+
+	return -1, fmt.Errorf("can't bucket %v", v)
+}
+
+type kv struct {
+	key   interface{}
+	value int // count
+}
+
+func anyBuckets(kvs []kv) ([]bucket, error) {
```

Comment:
Since anyBuckets has to match with anyBucketFn, and likewise for uniformBuckets and numericBucketFn, I'm wondering whether it's better to create a Buckets interface with an AddKey(key any, count int) method or similar? Open to debate though.

Reply:
mm yeah i like that

Reply:
went a different way in the end.. felt wrong somehow and i like what i ended up w; post-decision reasoning would be -
* not sure the interface makes much sense b/c there s so much overlap logic?
* and i think u just wanted me to merge those two which i did with "within"
* the add method specifically i didnt like b/c then i think it would make more sense to loop through the whole collection once and just bucket as i go, which i did initially some time ago, but this is a bit more modular ? 
* all of this deffo ties into our little discussion above also, b/c i wanted to keep my bounds struct at the base of all of this, but also i think overriding logic is nicer now, so hopefully u mind less

not gonna argue past this, i can just ditch everything and bucket as i go / interface on add; tho i do think that will make it more of a pain for us to specify different bucketing logic (at least in my head / but maybe i m imagining this wrong)